File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    
    public int GoodNodes(TreeNode root) {
        
        return Dfs(root, root.val);
    }
    public int Dfs (TreeNode node, int max){
        if (node is null) return 0;
        int Good;

        if (node.val >= max){
            Good = 1;
        }else {
            Good = 0;
        }
        max = Math.Max(node.val, max);

        Good += Dfs(node.left, max);
        Good += Dfs(node.right, max);

        return Good;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        #edge case 
        if not root:
            return True
        self.res = True
        self.dfs(root)

        return self.res

    def dfs(self, root):
        if root is None:
            return 0

        left = self.dfs(root.left)
        right = self.dfs(root.right)

        if abs(right - left ) > 1:
            self.res = False



        return 1 + max(left,right)
    

    #check if left and right are within one 
    #abs(left - right) > 1 